### PR TITLE
Using third-party source modules

### PR DESCRIPTION
**Pull request by alessandrogario**
This is a proposed solution for the libraries that are easy to convert; the remaining libraries (such as Boost, openssl, ...) would use the ExternalProject_add scripts from https://github.com/osquery/osquery/pull/5616

**Disclamer**: libraries are only inside the project (add_subdirectory) for convenience right now. We can move them to a find_package handler if we want (or even keep them separate).

Pro:
1. Single build system.
2. Not hard to convert the libraries
3. We don't have to store any binary package
4. Parallel builds are freaking fast, so the compilation time is **way** lower that a full source build with ExternalProject_add.
5. Cross dependency are handled in a sane manner (i.e. thirdparty_augeas will get libxml2 from our thirdparty_libxml2 target and so on).
6. Almost all the modules are directly attached to the upstream repository with a git submodule. This is super clean!
7. More importantly: **there is no compile-time system detection**. The configuration is frozen now, and we don't risk having a library auto-enable zlib support only because we forgot the library and the includes installed on the system.
8. All the compilation settings/flags are automatically inherited by osquery now!

### CODE REVIEW

**Comment by directionless** [comment]
I really like the goals, as represented in the pros here. 

I _think_ this is checking in the pinned submodule version, and the autoconf output? I'd love to see that documented somewhere. How to make a new one, how to upgrade one. Etc. (no need to block on docs, more a future ask about making this all smoother).

Are the submodule versions are pinned? I might have missed them in the diff, it's been awhile since I thought about submodules.

**Comment by alessandrogario** [comment]
> I really like the goals, as represented in the pros here.
> 
> I _think_ this is checking in the pinned submodule version, and the autoconf output? I'd love to see that documented somewhere. How to make a new one, how to upgrade one. Etc. (no need to block on docs, more a future ask about making this all smoother).
> 
> Are the submodule versions are pinned? I might have missed them in the diff, it's been awhile since I thought about submodules.

Thanks for looking into this @directionless !

All the submodules are pinned to the upstream release tags that correspond to the version expected by osquery. The only exception is sqlite because they don't have a git repository (and that's why the added lines here seem huge!)

Updating things is not really hard and it's simple as:
1. `cd third-party/zlib/src
2. `git fetch --tags && git checkout new tag`
3. `git log --name-status oldtag..newtag`
4. Take the generated file list and add/remove the files that have been changed/deleted/updated

You are correct, we only need the generated `config.h` file (making sure the enabled/disabled things match what we want). Sometimes they also generate things like `version.h` so we have to update that too

**Comment by directionless** [comment]
> Updating things is not really hard and it's simple as:

Please stick that into a developer doc somewhere. I know you and @theopolis are doing a lot, and I know future me is going to be lost.

**Comment by theopolis** [comment]
I like this implementation a lot. The generated configs and versions make sense to track within the osquery repo, SQLite does not. Can we hold the pinned SQLite code in a repository within the osquery organization and follow the same practice as the other dependencies?

**Comment by Smjert** [comment]
I think this is really good and solves the issues with separate build systems/build jobs parallelisation, beyond improving the ability to ensure exactly how they are built.

After solving some of the build issues the CI is reporting, I think we need to deal with a couple of small things:

1. Patching the third party code where necessary
2. Selectively checkout submodules (though CMake?), since as the CI shows not all submodules can be successfully checked out on Windows.
In general also it's a waste of time.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -42,11 +42,11 @@ jobs:
           $(EXTRA_CMAKE_ARGS)
           $(Build.SourcesDirectory)
 
-    - task: CMake@1
-      displayName: "Check code formatting"
-      inputs:
-        workingDirectory: $(Build.BinariesDirectory)/build
-        cmakeArgs: --build . --target format_check
+    #- task: CMake@1
```
Comment (theopolis) [inline azure-pipelines.yml]: Why?

File: cmake/options.cmake [hunk 2]
```diff
@@ -21,6 +21,9 @@ option(OSQUERY_NO_DEBUG_SYMBOLS "Whether to build without debug symbols or not,
 
 option(BUILD_TESTING "Whether to enable and build tests or not")
 
+# TODO: Revert this to 'facebook' before we merge!
+set(OSQUERY_THIRD_PARTY_SOURCE "source;pre-built;facebook" CACHE STRING "Sources used to acquire third-party dependencies")
```
Comment (theopolis) [inline cmake/options.cmake]: Does it make sense to update the other variables to follow this naming convention? `THIRD_PARTY_REPOSITORY_URL` becomes `OSQUERY_THIRD_PARTY_URL`?

Comment (theopolis) [reply cmake/options.cmake]: Can `facebook` become `legacy` or similar?

File: azure-pipelines.yml (hunk 1 above)
Comment (alessandrogario) [reply azure-pipelines.yml]: Everything I changed in the Azure pipelines file is temporary; I'll restore the original one in the final commit

File: cmake/options.cmake (hunk 2 above)
Comment (alessandrogario) [reply cmake/options.cmake]: I didn't call it 'legacy' as it's the main source for Buck libraries (and it's still actively used), and my aim is to stay 100% compatible

**Comment by alessandrogario** [comment]
Closing, and merging this with https://github.com/osquery/osquery/pull/5616 directly inside the branch

**Pull request closed without merging**